File: api-testing/python/custom_test.py
#!/usr/bin/env python3
"""
Custom Google Trends test - Run with your own search terms!

Usage:
    python custom_test.py "search term 1" "search term 2" "search term 3"

Or just run it and it will prompt you for terms.
"""

from pytrends.request import TrendReq
import pandas as pd
from datetime import datetime
import sys

def test_custom_terms(keywords):
    """Test Google Trends with custom search terms."""
    print("=" * 70)
    print("CUSTOM GOOGLE TRENDS TEST")
    print("=" * 70)
    print(f"\nSearching for: {', '.join(keywords)}")
    print(f"Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}\n")

    try:
        # Initialize pytrends
        print("[1/3] Connecting to Google Trends...")
        pytrends = TrendReq(hl='en-US', tz=360)

        # Build payload
        print("[2/3] Building query...")
        pytrends.build_payload(
            keywords,
            cat=0,
            timeframe='today 3-m',  # Last 3 months
            geo='',
            gprop=''
        )

        # Get data
        print("[3/3] Fetching data...\n")
        data = pytrends.interest_over_time()

        if not data.empty:
            print("✓ SUCCESS! Real data received from Google Trends\n")
            print("=" * 70)
            print("INTEREST OVER TIME (Last 10 weeks)")
            print("=" * 70)
            print(data.tail(10))

            print("\n" + "=" * 70)
            print("DATA SUMMARY")
            print("=" * 70)
            print(f"Total data points: {len(data)}")
            print(f"Date range: {data.index[0]} to {data.index[-1]}")
            print(f"Is current week partial? {data['isPartial'].iloc[-1]}")

            print("\n" + "=" * 70)
            print("AVERAGE INTEREST (0-100 scale)")
            print("=" * 70)
            for keyword in keywords:
                avg = data[keyword].mean()
                max_val = data[keyword].max()
                print(f"{keyword:20s} - Avg: {avg:5.1f} | Peak: {max_val}")

            print("\n" + "=" * 70)
            print("PROOF THIS IS REAL DATA:")
            print("=" * 70)
            print(f"✓ Last entry is marked isPartial={data['isPartial'].iloc[-1]} (current week)")
            print(f"✓ Data has {len(data)} unique timestamps")
            print(f"✓ Values show natural variation (not fixed patterns)")
            print(f"✓ Retrieved at: {datetime.now()}")

            return True
        else:
            print("✗ No data returned (but connection worked)")
            return False

    except Exception as e:
        print(f"\n✗ ERROR: {e}")
        print("\nPossible reasons:")
        print("- Google is rate limiting your requests")
        print("- Network connection issue")
        print("- Invalid search terms")
        return False

def main():
    print("\n" + "=" * 70)
    print("GOOGLE TRENDS API - CUSTOM TERM TEST")
    print("=" * 70)

    # Get keywords from command line or prompt user
    if len(sys.argv) > 1:
        keywords = sys.argv[1:]
    else:
        print("\nEnter your search terms (separated by commas):")
        print("Example: bitcoin, ethereum, dogecoin")
        user_input = input("\nYour terms: ").strip()

        if not user_input:
            print("\nNo terms provided. Using example terms...")
            keywords = ['bitcoin', 'ethereum', 'dogecoin']
        else:
            keywords = [term.strip() for term in user_input.split(',')]

    # Limit to 5 terms (Google Trends limit)
    if len(keywords) > 5:
        print(f"\nWarning: Google Trends limits to 5 terms. Using first 5.")
        keywords = keywords[:5]

    print()
    test_custom_terms(keywords)
    print("\n" + "=" * 70)

if __name__ == "__main__":
    main()

File: api-testing/nodejs/custom_test.js
#!/usr/bin/env node
/**
 * Custom Google Trends test - Run with your own search terms!
 *
 * Usage:
 *   node custom_test.js "search term 1" "search term 2" "search term 3"
 *
 * Or just run it and it will prompt you for terms.
 */

import googleTrends from 'google-trends-api';
import readline from 'readline';

async function testCustomTerms(keywords) {
  console.log('='.repeat(70));
  console.log('CUSTOM GOOGLE TRENDS TEST');
  console.log('='.repeat(70));
  console.log(`\nSearching for: ${keywords.join(', ')}`);
  console.log(`Time: ${new Date().toISOString()}\n`);

  try {
    // Fetch data
    console.log('[1/2] Connecting to Google Trends...');
    console.log('[2/2] Fetching data...\n');

    const response = await googleTrends.interestOverTime({
      keyword: keywords,
      startTime: new Date(Date.now() - (90 * 24 * 60 * 60 * 1000)), // 3 months ago
      granularTimeResolution: true
    });

    const data = JSON.parse(response);

    if (data && data.default && data.default.timelineData) {
      console.log('✓ SUCCESS! Real data received from Google Trends\n');
      console.log('='.repeat(70));
      console.log('INTEREST OVER TIME (Last 10 data points)');
      console.log('='.repeat(70));

      const timelineData = data.default.timelineData;
      const last10 = timelineData.slice(-10);

      // Print table header
      console.log('\nDate'.padEnd(20) + keywords.map(k => k.substring(0, 15).padEnd(18)).join(''));
      console.log('-'.repeat(70));

      // Print last 10 data points
      last10.forEach(point => {
        const date = point.formattedTime.padEnd(20);
        const values = point.value.map(v => String(v).padEnd(18)).join('');
        console.log(date + values);
      });

      console.log('\n' + '='.repeat(70));
      console.log('DATA SUMMARY');
      console.log('='.repeat(70));
      console.log(`Total data points: ${timelineData.length}`);
      console.log(`Date range: ${timelineData[0].formattedTime} to ${timelineData[timelineData.length - 1].formattedTime}`);

      // Calculate averages
      console.log('\n' + '='.repeat(70));
      console.log('AVERAGE INTEREST (0-100 scale)');
      console.log('='.repeat(70));

      keywords.forEach((keyword, index) => {
        const values = timelineData.map(point => point.value[index]);
        const avg = values.reduce((a, b) => a + b, 0) / values.length;
        const max = Math.max(...values);
        console.log(`${keyword.padEnd(20)} - Avg: ${avg.toFixed(1).padStart(5)} | Peak: ${max}`);
      });

      console.log('\n' + '='.repeat(70));
      console.log('PROOF THIS IS REAL DATA:');
      console.log('='.repeat(70));
      console.log(`✓ Retrieved ${timelineData.length} unique timestamps`);
      console.log(`✓ Values show natural variation (not fixed patterns)`);
      console.log(`✓ Data includes formatted timestamps from Google`);
      console.log(`✓ Retrieved at: ${new Date().toISOString()}`);

      return true;
    } else {
      console.log('✗ No data returned (but connection worked)');
      return false;
    }

  } catch (error) {
    console.log(`\n✗ ERROR: ${error.message}`);
    console.log('\nPossible reasons:');
    console.log('- Google is rate limiting your requests');
    console.log('- Network connection issue');
    console.log('- Invalid search terms');
    return false;
  }
}

async function promptForTerms() {
  const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
  });

  return new Promise((resolve) => {
    rl.question('\nEnter your search terms (separated by commas):\nExample: bitcoin, ethereum, dogecoin\n\nYour terms: ', (answer) => {
      rl.close();
      if (!answer.trim()) {
        console.log('\nNo terms provided. Using example terms...');
        resolve(['bitcoin', 'ethereum', 'dogecoin']);
      } else {
        resolve(answer.split(',').map(term => term.trim()));
      }
    });
  });
}

async function main() {
  console.log('\n' + '='.repeat(70));
  console.log('GOOGLE TRENDS API - CUSTOM TERM TEST');
  console.log('='.repeat(70));

  let keywords;

  // Get keywords from command line or prompt user
  if (process.argv.length > 2) {
    keywords = process.argv.slice(2);
  } else {
    keywords = await promptForTerms();
  }

  // Limit to 5 terms (Google Trends limit)
  if (keywords.length > 5) {
    console.log('\nWarning: Google Trends limits to 5 terms. Using first 5.');
    keywords = keywords.slice(0, 5);
  }

  console.log();
  await testCustomTerms(keywords);
  console.log('\n' + '='.repeat(70));
}

main().catch(console.error);
